README.md
## Sagemode Jutsu: Simple and Effective OSINT Username Search Tool

<p align="center">
 <img src="https://github.com/senran101604/sagemode/blob/master/gif/example_usage.gif" width="65%"
</p>

## Introduction
Sagemode Jutsu is a straightforward and efficient Open Source Intelligence
(OSINT) tool created with simplicity in mind. It allows users to search for
specific usernames across various online platforms, aiding digital investigators
and cybersecurity enthusiasts in their research.


## Installation
**Clone the Repo**
```console
git clone https://github.com/senran101604/sagemode
```
**change directory to sagemode**
```console
cd sagemode
```
**Install Requirements**

```console
python3 -m pip install -r requirements.txt
```

## Reminder
**Please regularly update SageMode for better user experience:**
```console
❯ python3 sagemode.py -U
```

## Key Features
- **Easy to Use:** A user-friendly tool designed for effortless username
  hunting without unnecessary complications.
- **Real-time Feedback:** Receive instant feedback on positive search results,
  providing insight into the username's online presence.
- **Curated List** of sites for more manageable investigation.
  Users are encouraged to submit pull requests if they fix issues or add useful features.

## How to Use
```console
❯ python3 .\sagemode.py --help
usage: sagemode.py [-h] [-f] [-v] [-U] [username]

Sagemode Jutsu: Unleash Your Inner Ninja

positional arguments:
  username       username to search for

optional arguments:
  -h, --help     show this help message and exit
  -f, --found    output only found sites
  -v, --version  Print Sagemode Version
  -U, --update   update Sagemode

```
#### To search for a <target_username> example
```console
$python sagemode.py <target_username>
```

1. Enter the target username.
2. Sagemode Jutsu will deliver real-time results, highlighting where the
   username has been found online.


## Use Cases
- **For Fun**: Explore the online presence of friends, celebrities, or
  fictional characters for entertainment purposes.
- **Digital Investigations:** Gather information about individuals across online
  platforms for investigative purposes.
- **Cybersecurity:** Identify potential security concerns by detecting
  suspicious usernames linked to malicious activities.
- **Online Reputation Management:** Monitor online presence and address
  potential identity-related issues.
- **Research and Analysis:** Gather data for social media analysis, market
  research, or competitive analysis.


## Disclaimer
* Sagemode Jutsu is intended for ethical and legal use. Users are responsible for
  adhering to applicable laws and regulations while utilizing the tool.
* Please be reminded that Sagemode is still in development and might return
  false results. Sites that are most prone to false positives are Facebook and
  Khan Academy.
* I'm in the process of learning and improving. If you find or fix issues or
  have feature suggestions, your pull requests are welcome.


## TODO
- [X] Reduce soft 404 error responses.
- [ ] Add Tests, Sites, etc...
- [ ] Cleaner Code :)


## Contributions
Contributions and feedback are appreciated! Feel free to submit issues,
suggestions, or enhancements via GitHub.

-------------------------------------------------------------------------------


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.1.2] - 2023-11-12
### Added
- Better explanation of codes.
- This CHANGELOG file to hopefully help track this project's progress.
- An feature to update Sagemode.
### Changed
- Spinner for rich's status.
- Made the Notify class as a helper class instead of inheriting from it.
- Methods of Notify class from instance methods to a static methods.
- Rely on rich library instead of the color() function from accessories module that
  has been removed.
### Fixed
- Error pretty printing the banner in windows command prompt.
### Removed
- color() function in accessories module.

## [1.1.3] - 2023-11-18
### Added
- User-Agents list in sites.py to avoid being blocked by the servers
- A way to output the status code in notify.not_found() method.
- More sites and more error checkings for false positives.
### Fixed
- Soft 404 errors (reduced false positive responses of sites)


sites.py
sites = {
    # Social Media Platforms
    "Facebook": "https://www.facebook.com/{}",
    "Twitter": "https://twitter.com/{}",
    "Instagram": "https://www.instagram.com/{}",
    "YouTube": "https://www.youtube.com/{}",
    "Snapchat": "https://www.snapchat.com/add/{}",
    "TikTok": "https://www.tiktok.com/@{}",
    "Reddit": "https://www.reddit.com/user/{}",
    # E-commerce & Marketplace
    "eBay": "https://www.ebay.com/usr/{}",
    "Fiverr": "https://www.fiverr.com/{}",
    "Amazon": "https://www.amazon.com/gp/profile/amzn1.account.{}",
    # Educational Platforms
    "Khan Academy": "https://www.khanacademy.org/profile/{}",
    "Coursera": "https://www.coursera.org/user/{}",
    # Professional & Business Networks
    "Crunchbase": "https://www.crunchbase.com/person/{}",
    "GitHub Gist": "https://gist.github.com/{}",
    "GitHub": "https://www.github.com/{}",
    "Gitstar Ranking": "https://gitstar-ranking.com/{}",
    "GitLab": "https://gitlab.com/{}",
    "Bitbucket": "https://bitbucket.org/{}",
    "Behance": "https://www.behance.net/{}",
    "Dribbble": "https://dribbble.com/{}",
    "Stack Overflow": "https://stackoverflow.com/users/{}",
    # Creative & Multimedia Platforms
    "Pinterest": "https://www.pinterest.com/{}",
    "DeviantArt": "https://{}.deviantart.com",
    "Patreon": "https://www.patreon.com/{}",
    "Flickr": "https://www.flickr.com/people/{}",
    "Vimeo": "https://vimeo.com/{}",
    "SoundCloud": "https://soundcloud.com/{}",
    # Blogging & Writing Platforms
    "Blogger": "https://{}.blogspot.com",
    "Tumblr": "https://{}.tumblr.com/",
    "Medium": "https://medium.com/@{}",
    "Wix": "https://{}.wixsite.com/website",
    "Weebly": "https://{}.weebly.com",
    "Wellfound": "https://wellfound.com/u/{}",
    "WordPress": "https://{}.wordpress.com",
    "WordPressOrg": "https://profiles.wordpress.org/{}",
    "Weblate": "https://hosted.weblate.org/user/{}",
    "LiveJournal": "https://{}.livejournal.com",
    "BuzzFeed": "https://buzzfeed.com/{}",
    # Other Platforms
    "Houzz": "https://houzz.com/user/{}",
    "Gravatar": "https://gravatar.com/{}",
    "SlideShare": "https://www.slideshare.net/{}",
    "HubPages": "https://hubpages.com/@{}",
    "Wikipedia": "https://en.wikipedia.org/wiki/User:{}",
    "DailyMotion": "https://www.dailymotion.com/{}",
    "Vine": "https://vine.co/{}",
    "Mixcloud": "https://www.mixcloud.com/{}/",
    "Patreon": "https://www.patreon.com/{}",
    "Quora": "https://www.quora.com/profile/{}",
    "Wattpad": "https://www.wattpad.com/user/{}",
    # Programming & Coding Platforms
    "LeetCode": "https://leetcode.com/{}",
    "Coderbyte": "https://www.coderbyte.com/profile/{}",
    "Codecademy": "https://www.codecademy.com/profiles/{}",
    # Gaming & Streaming Platforms
    "AQW": "https://account.aq.com/CharPage?id={}",
    "Steam": "https://steamcommunity.com/id/{}",
    "SteamGroup": "https://steamcommunity.com/groups/{}",
    "SourceForge": "https://sourceforge.net/u/{}",
    "Twitch": "https://www.twitch.tv/{}",
    "Nintendo Life": "https://www.nintendolife.com/users/{}",
    "GOG": "https://www.gog.com/u/{}",
    "Spotify": "https://open.spotify.com/user/{}",
    # Additional Platforms
    "PyPi": "https://pypi.org/user/{}",
    "SlideShare": "https://www.slideshare.net/{}",
    "Slides": "https://slides.com/{}",
    "F6S": "https://www.f6s.com/{}",
    "Dribbble": "https://dribbble.com/{}",
    "Gravatar": "https://en.gravatar.com/{}",
    "SpeakerDeck": "https://speakerdeck.com/{}",
    "Behance": "https://www.behance.net/{}",
    "Academia.edu": "https://{}.academia.edu",
    "Vimeo": "https://vimeo.com/{}",
    "Keybase": "https://keybase.io/{}",
    "Bitbucket": "https://bitbucket.org/{}",
    "GitLab": "https://gitlab.com/{}",
    "Goodreads": "https://www.goodreads.com/{}",
    "Last.fm": "https://www.last.fm/user/{}",
    "ReverbNation": "https://www.reverbnation.com/{}",
    "Letterboxd": "https://letterboxd.com/{}",
    "Bandcamp": "https://bandcamp.com/{}",
    "Vine": "https://vine.co/{}",
    "Quora": "https://www.quora.com/profile/{}",
    "Medium": "https://medium.com/@{}",
    "Mixcloud": "https://www.mixcloud.com/{}",
    "ProductHunt": "https://www.producthunt.com/@{}",
    "Steam": "https://steamcommunity.com/id/{}",
    "Wikipedia": "https://en.wikipedia.org/wiki/User:{}",
    "Reddit": "https://www.reddit.com/user/{}",
    "Slack": "https://{}.slack.com",
    "MySpace": "https://www.myspace.com/{}",
    "MyAnimeList": "https://myanimelist.net/profile/{}",
    "BuyMeACoffee": "https://www.buymeacoffee.com/{}",
    "Freelancer": "https://www.freelancer.com/u/{}",
    "Mapify": "https://mapify.travel/{}",
    "Pastebin": "https://pastebin.com/u/{}",
    "Mydramalist": "https://www.mydramalist.com/profile/{}",
    "Scribd": "https://www.scribd.com/{}",
    "Lichess": "https://lichess.org/@/{}",
    "Chess": "https://www.chess.com/member/{}",
    "Osu!": "https://osu.ppy.sh/users/{}",
    "Fixya": "https://www.fixya.com/users/{}",
    "Issuu": "https://issuu.com/{}",
    "Venmo": "https://account.venmo.com/u/{}",
    "ThemeForest": "https://themeforest.net/user/{}",
    "TradingView": "https://www.tradingview.com/u/{}",
    "TETR.IO": "https://ch.tetr.io/u/{}",
    "Scratch": "https://scratch.mit.edu/users/{}",
    "RuneScape": "https://apps.runescape.com/runemetrics/app/overview/player/{}",
    "Rumble": "https://rumble.com/user/{}",
    "Roblox": "https://www.roblox.com/user.aspx?username={}",
    "Replit.com": "https://replit.com/@{}",
    "Pokemon Showdown": "https://pokemonshowdown.com/users/{}",
    "OpenStreetMap": "https://www.openstreetmap.org/user/{}",
    "Monkeytype": "https://monkeytype.com/profile/{}",
    "Linktree": "https://linktr.ee/{}",
    "Kongregate": "https://www.kongregate.com/accounts/{}",
    "Instructables": "https://www.instructables.com/member/{}",
    "Imgur": "https://imgur.com/user/{}",
    "HackerRank": "https://hackerrank.com/{}",
    "GaiaOnline": "https://www.gaiaonline.com/profiles/{}",
    "Flipboard": "https://flipboard.com/{}",
    "Disqus": "https://disqus.com/by/{}",
    "Audiojungle": "https://audiojungle.net/user/{}",
    "Apple Discussions": "https://discussions.apple.com/profile/{}",
    "Anilist": "https://anilist.co/user/{}",
    "note": "https://note.com/{}",
    "Minds": "https://www.minds.com/{}",
    "uID": "https://uid.me/{}",
    "Interpals": "https://www.interpals.net/{}",
    "Geocaching": "https://www.geocaching.com/p/default.aspx?u={}",
    "Fur Affinity": "https://www.furaffinity.net/user/{}",
    "devRant": "https://devrant.com/users/{}",
    "freeCodeCamp": "https://www.freecodecamp.org/{}",
    "Couchsurfing!": "https://www.couchsurfing.com/people/{}",
    "YouPic": "https://youpic.com/photographer/{}",
    "Unsplash": "https://unsplash.com/@{}",
    "Ultimate-Guitar": "https://ultimate-guitar.com/u/{}",
    "TryHackMe": "https://tryhackme.com/p/{}",
    "Trello": "https://trello.com/{}",
    "Trakt": "https://www.trakt.tv/users/{}",
    "Periscope": "https://www.periscope.tv/{}",
    "Polarsteps": "https://polarsteps.com/{}",
    "Naver": "https://blog.naver.com/{}",
    "Hackaday": "https://hackaday.io/{}",
    "Choice Community": "https://choice.community/u/{}/summary",
    "Clapper": "https://clapperapp.com/{}",
    "Bikemap": "https://www.bikemap.net/en/u/{}/routes/created/",
    "BioHacking": "https://forum.dangerousthings.com/u/{}",
}

# indicators for false positive 200 responses
soft404_indicators = [
    "This profile could not be found",
    "Sorry, this user was not found",
    "Page Not Found",
    "the profile was either removed ",
    "The specified profile could not be found",
    "doesn&apos;t&nbsp;exist",
    "This page doesn't exist",
    "404 Not Found",
    "Sorry, nobody on Reddit goes by that name",
    "The person may have been banned or the username is incorrect.",
    "Some error occured while loading page for you. Please try again",
    "No longer a registered user",
    "No user ID specified or user does not exist!",
    "user does not exist",
    "User not found.",
    "An error was encountered while processing your request",
    "Sorry, the page you were looking for doesn’t exist",
    "Wikipedia does not have a",
    "We could not find the page you were looking for, so we found something to make you laugh to make up for it",
    "Not Found!",
    '"statusMsg":"","needFix"',
]

user_agents = [
    # Chrome 115.0 on macOS
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36",
    # Chrome 114.0 on Windows
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36",
    # Chrome 118.0 on Windows
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/118.0.0.0 Safari/537.36",
    # Chrome 119.0 on Windows
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36",
    # Firefox 116.0 on Windows
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/116.0",
    # Firefox 115.0 on Windows
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/115.0",
    # Firefox 119.0 on Windows
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/119.0",
    # Chrome 114.0 on macOS
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36",
    #  Chrome 115.0 on Linux
    "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36",
    # Chrome 116.0 on Windows
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36",
    # Firefox 115.0 on Linux
    "Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/115.0",
    # Firefox 116.0 on Linux
    "Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/116.0",
    # Edge 115.0 on Windows
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36 Edg/115.0.1901.188",
]

if __name__ == "__main__":
    from rich import print as rprint

    for site, url in sites.items():
        rprint(f"[bright_green]{site}: [blue]{url}")

    rprint(f"\n[magenta]{len(sites)} Sites")


accessories.py
"""
accessory module used to design and color the output text for pretty terminal
output.
"""

ascii_art = """
       ____                __  ___        __
      / __/__ ____ ____   /  |/  /__  ___/ /__
     _\ \/ _ `/ _ `/ -_) / /|_/ / _ \/ _  / -_)
    /___/\_,_/\_, /\__/ /_/  /_/\___/\_,_/\__/
             /___/

                  @@@%%%%%@@@
              @%##`````@`````##&@
            @##````````@````````##@
          @%#`````````@@@`````````#%@
          &#``````````@@@``````````#&
         @#````@@@@@@@@@@@@@@@@@````#@
         @%@``@@@@@@@@@@@@@@@@@@@``@%@
         @%@```@@@@@@@@@@@@@@@@@```#%@
         @@# `````````@@@``````````#@@
          &#``````````@@@``````````#&
           @##`````````@`````````##@
             @##```````@``````###@
                @@#````@````#@@
                  @@@%%%%%@@@
"""


def start(banner: str, delay=0.001):
    """
    Parameters:
        banner
        delay=0.2
    """
    # import random
    from time import sleep
    from rich import print as rprint

    for line in banner.split("\n"):
        for character in line:
            if character in ["#", "@", "%", "&"]:
                rprint(f"[yellow]{character}", end="", flush=True)
            else:
                rprint(f"[bright_red]{character}", end="", flush=True)
        print()


class Notify:
    "A helper class for notifications of Sagemode process"

    @staticmethod
    def start(username: str, number_of_sites) -> str:
        start(ascii_art, delay=0.1)
        if username or sites is not None:
            return f"[yellow][[bright_red]*[yellow][yellow]] [bright_blue]Searching {number_of_sites} sites for target: [bright_yellow]{username}"

    # notify the user how many sites the username has been found
    @staticmethod
    def positive_res(username: str, count) -> str:
        return f"\n[yellow][[bright_red]+[yellow]][bright_green] Found [bright_red]{username} [bright_green]on [bright_magenta]{count}[bright_green] sites"

    # notify the user where the result is stored
    @staticmethod
    def stored_result(result_file: str) -> str:
        return f"[bright_green][[yellow]@[bright_green]] [orange3]Results stored in: [bright_green]{result_file}\n"

    @staticmethod
    def not_found(site: str, status_code="") -> str:
        if status_code:
            return f"[black][[red]-[black]] [blue]{site}: [yellow]Not Found! {status_code}"
        return f"[black][[red]-[black]] [blue]{site}: [yellow]Not Found!"

    @staticmethod
    def found(site: str, url: str) -> str:
        return f"[red][[green]+[red]] [green]{site}: [blue]{url}"

    @staticmethod
    def update(local_version: str, remote_version: str) -> str:
        return (
            "[red][[bright_red]![red]] [yellow]Update Available!\n[/yellow]"
            + f"[red][[yellow]![red]] [bright_yellow]You are running Version: [bright_green]{local_version}\n"
            + f"[red][[/red][yellow]![red]][bright_yellow] New Version Available: [bright_green]{remote_version}"
        )

    @staticmethod
    def update_error(error: str) -> str:
        return f"[bright_red][[bright_red]![bright_red]] [bright_yellow]A problem occured while checking for an update: [bright_red]{error}"

    @staticmethod
    def version(version: str) -> str:
        return f"[bright_yellow]Sagemode [bright_red]{version}"

    def exception(site, error):
            return f"[black][[red]![black]] [blue]{site}: [bright_red]{error}..."


if __name__ == "__main__":
    start(ascii_art)


sagemode.py
#! /usr/bin/env python3
"""
Sagemode: Track and Unveil Online identities across social media platforms.
"""
import os
import re
import datetime
import subprocess
import threading
import random
import requests
from argparse import ArgumentParser
from rich.console import Console
from bs4 import BeautifulSoup

from accessories import Notify
from sites import sites, soft404_indicators, user_agents


__version__ = "1.1.3"


class Sagemode:
    def __init__(self, username: str, found_only=False):
        self.console = Console()
        self.notify = Notify
        self.positive_count = 0
        self.username = username
        self.result_file = os.path.join("data", f"{self.username}.txt")
        self.found_only = found_only

    # this function checks if the url not a false positive result, return false
    def is_soft404(self, html_response: str) -> bool:
        # this is for checking the title bar of the page
        soup = BeautifulSoup(html_response, "html.parser")
        page_title = soup.title.string.strip() if soup.title else ""

        # I know this is kinda messy solution but it currently solve.. reduce the problem
        # in soft404 responses (false positives)
        for error_indicator in soft404_indicators:
            if (
                # check if the error indicator is in the html string response
                error_indicator.lower() in html_response.lower()
                # check for the title bar of the page if there are anyi error_indicator
                or error_indicator.lower() in page_title.lower()
                # Specific check sites, since positive result will have the username in the title bar.
                or page_title.lower() == "instagram"
                # patreon's removed user
                or page_title.lower() == "patreon logo"
                or "sign in" in page_title.lower()
            ):
                return True
        return False

    def check_site(self, site: str, url: str, headers):
        url = url.format(self.username)
        # we need headers to avoid being blocked by requesting the website 403 error
        try:
            with requests.Session() as session:
                response = session.get(url, headers=headers)
                # Raises an HTTPError for bad responses
            # further check to reduce false positive results
            if (
                response.status_code == 200
                and self.username.lower() in response.text.lower()
                and not self.is_soft404(response.text)
            ):
                # to prevent multiple threads from accessing/modifying the positive
                # counts simultaneously and prevent race conditions.
                with threading.Lock():
                    self.positive_count += 1
                self.console.print(self.notify.found(site, url))
                with open(self.result_file, "a") as f:
                    f.write(f"{url}\n")
            # the site reurned 404 (user not found)
            else:
                if not self.found_only:
                    self.console.print(self.notify.not_found(site))
        except Exception as  e:
            self.notify.exception(site, e)

    def start(self):
        """
        Start the search.
        """
        self.console.print(self.notify.start(self.username, len(sites)))

        current_datetime = datetime.datetime.now()
        date = current_datetime.strftime("%m/%d/%Y")
        time = current_datetime.strftime("%I:%M %p")
        headers = {"User-Agent": random.choice(user_agents)}

        with open(self.result_file, "a") as file:
            file.write(f"\n\n{29*'#'} {date}, {time} {29*'#'}\n\n")

        # keep track of thread objects.
        threads = []

        try:
            with self.console.status(
                f"[*] Searching for target: {self.username}", spinner="bouncingBall"
            ):
                for site, url in sites.items():
                    # creates a new thread object
                    thread = threading.Thread(target=self.check_site, args=(site, url, headers))
                    # track the thread objects by storing it in the assigned threads list.
                    threads.append(thread)
                    # initiate the execution of the thread
                    thread.start()
                for thread in threads:
                    # waits for each thread to finish before proceeding.
                    # to avoid output problems and maintain desired order of executions
                    thread.join()

            # notify how many sites the username has been found
            self.console.print(
                self.notify.positive_res(self.username, self.positive_count)
            )
            # notify where the result is stored
            self.console.print(self.notify.stored_result(self.result_file))

        except Exception:
            self.console.print_exception()

    def check_for_update(self):
        try:
            r = requests.get(
                "https://raw.githubusercontent.com/senran101604/sagemode/master/sagemode.py"
            )

            remote_version = str(re.findall('__version__ = "(.*)"', r.text)[0])
            local_version = __version__

            if remote_version != local_version:
                self.console.print(self.notify.update(local_version, remote_version))

        except Exception as error:
            self.console.print(self.notify.update_error(error))

    def do_update(self):
        repo_dir = os.path.dirname(os.path.realpath(__file__))
        # ensure we're performing git command in the local git repo directory
        os.chdir(repo_dir)
        subprocess.run(["git", "pull"])


def main():
    # create data folder if it doesn't exist
    if not os.path.exists("data"):
        os.mkdir("data")

    parser = ArgumentParser(description="Sagemode Jutsu: Unleash Your Inner Ninja")
    parser.add_argument(
        "username", help="username to search for", action="store", nargs="?"
    )
    parser.add_argument(
        "-f",
        "--found",
        dest="found",
        help="output only found sites",
        action="store_true",
    )
    parser.add_argument(
        "-v",
        "--version",
        dest="show_version",
        help="Print Sagemode Version",
        action="store_true",
    )
    parser.add_argument(
        "-U", "--update", dest="do_update", help="update Sagemode", action="store_true"
    )
    args = parser.parse_args()

    sage = Sagemode(args.username, found_only=args.found)

    # check if the username arguemnt is given then start searching, this will ensure
    # that we can use the --version flag without supplying the username.
    if args.username is not None:
        sage.start()
    if args.show_version:
        sage.console.rule(sage.notify.version(__version__))
    if args.do_update:
        sage.do_update()

    # always check for update in every run
    sage.check_for_update()


if __name__ == "__main__":
    main()
